stdlib/public/core/StringCharacterView.swift

```diff
@@ -76,15 +93,33 @@ extension String: BidirectionalCollection {
   ///   `startIndex`.
   /// - Returns: The index value immediately before `i`.
   public func index(before i: Index) -> Index {
-    _precondition(i > startIndex, "String index is out of bounds")
+    let i = _guts.ensureMatchingEncoding(i)
+    _precondition(i > startIndex && i <= endIndex,
+      "String index is out of bounds")
+    let r = _uncheckedIndex(before: _guts.scalarAlign(i))
+    return _guts.markEncoding(r)
+  }
+
+  /// A version of `index(before:)` that assumes that the given index:
+  ///
+  /// - has the right encoding,
+  /// - is within bounds, and
+  /// - is scalar aligned.
+  ///
+  /// It does not mark the encoding of the returned index.
+  internal func _uncheckedIndex(before i: Index) -> Index {
+    _internalInvariant(_guts.hasMatchingEncoding(i))
+    _internalInvariant(i > startIndex && i <= endIndex)
+    _internalInvariant(i._isScalarAligned)
 
     // TODO: known-ASCII fast path, single-scalar-grapheme fast path, etc.
-    let i = _guts.scalarAlign(i)
     let stride = _characterStride(endingAt: i)
     let priorOffset = i._encodedOffset &- stride
-    return Index(
-      encodedOffset: priorOffset, characterStride: stride)._scalarAligned
+
+    let r = Index(encodedOffset: priorOffset, characterStride: stride)
+    return r._scalarAligned
```

Comment:
`r._graphemeAligned`, which would set both

Reply:
Would `_characterAligned` be a better name?

Reply:
"character" is wildly ambiguous, I feel like. `Swift.Character` is a name for grapheme clusters, but for internal code/names, I think it's better to be specific

Reply:
In the context of the Swift stdlib, I think the definition of a character cannot possibly be clearer. `String` is a `Collection` of `Character` values; these routines help identify the boundaries between `Character`s. The words “string”, “collection” and “character” all have ambiguous meanings outside of this context, but inside of this codebase, we narrowed things down to extremely precise technical definitions.

If we must ignore the context this code is in (why?), then “grapheme” is just as ambiguous as character is — e.g., the “ly” in my name is a single letter, and it is considered one grapheme according to at least [some](https://eric.ed.gov/?id=ED072689) linguists. Unicode’s glossary has a [non-technical definition of grapheme](https://www.unicode.org/glossary/#grapheme) that doesn’t clear things up at all. The mere fact that it is a different name than `Character` invites confusion — if I came across it, I would suspect that it means something different.

Unicode only has unambiguous definitions for “grapheme cluster” and “extended grapheme cluster”, and iiuc `Character` implements the latter. So if we want to avoid confusion for people working on stdlib code who aren’t familiar with Swift’s `Character` type (who are these people?), then we should probably go with `_extendedGraphemeClusterAligned`.

Reply:
I think you're missing the point. Stdlib API is formally layered on top of an understanding of Unicode semantics, and "character" in Unicode has many definitions. If you want to cite Union's Glossary, then read [character](https://www.unicode.org/glossary/#character). "Extended" is meaningless except to contrast with legacy clusters. "Grapheme cluster" is the more precise name, I only mention "grapheme" because it clearly contrasts with scalars and code units.